cpp/modules/math.gl/core/src/core.h

```diff
@@ -595,13 +610,93 @@ auto operator<<(std::ostream &os, const Vector3<coord> &v) -> std::ostream & {
 // Vector4 implementation //
 ////////////////////////////
 
+template <typename coord>
+auto Vector4<coord>::length() const -> coord {
+  return static_cast<coord>(sqrt(x * x + y * y + z * z + w * w));
+}
+
+template <typename coord>
+auto Vector4<coord>::length2() const -> coord {
+  return static_cast<coord>(x * x + y * y + z * z + w * w);
+}
+
+template <typename coord>
+auto Vector4<coord>::toVector3() const -> Vector3<coord> {
+  return Vector3<coord>(this->x, this->y, this->z);
+}
+
 template <typename coord>
 auto Vector4<coord>::transform(const Matrix4<coord> &m) const -> Vector4<coord> {
   return Vector4<coord>{
       m(0, 0) * x + m(0, 1) * y + m(0, 2) * z + m(0, 3) * w, m(1, 0) * x + m(1, 1) * y + m(1, 2) * z + m(1, 3) * w,
       m(2, 0) * x + m(2, 1) * y + m(2, 2) * z + m(2, 3) * w, m(3, 0) * x + m(3, 1) * y + m(3, 2) * z + m(3, 3) * w};
 }
 
+template <typename coord>
+void Vector4<coord>::normalize() {
+  coord length = length();
+  if (length <= static_cast<coord>(0)) throw std::runtime_error("normalize called on zero length vector");
+  x /= length;
+  y /= length;
+  z /= length;
+  w /= length;
+}
+
+template <typename coord>
+auto vectorProduct(const Vector4<coord> &v1, const Vector4<coord> &v2, const Vector4<coord> &v3) -> Vector4<coord> {
```

Comment:
I believe the initial method signature with 2 vectors was supposed to represent a cross product, this seems like https://en.wikipedia.org/wiki/Triple_product? Not sure where this may be useful

We also may want to rename `vectorProduct` -> `crossProduct` and `elementProduct` -> `dotProduct` if that's what they're supposed to represent, as it should be a more common terminology

Reply:
Agreed. I am not sure these make sense for Vector4's though as those are used to represent homogeneous coordinates (projective geometry) rather than 4 dimensional positions, suggest dropping this overload for now.

Reply:
Dropped